src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -583,22 +583,28 @@ private PollingResult compareRemoteRevisionWithImpl(Job<?, ?> project, Launcher
                     for (BranchSpec branchSpec : getBranches()) {
                         for (Entry<String, ObjectId> entry : heads.entrySet()) {
                             final String head = entry.getKey();
-                            String name;
                             // head is "refs/(heads|tags)/branchName
-                            if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
-                            else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
-                            else name = remote + "/" + head;
 
-                            if (!branchSpec.matches(name, environment)) continue;
+                            // first, check the a canonical git reference is configured
+                            if (!branchSpec.matches(head, environment)) {
+
+                                // convert head `refs/*` into shortcut notation `remote/branch`
+                                String name = head;
+                                if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
+                                else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
```

Comment:
So you suggest to match `refs/*/` ? I never heard about refs/notes usage with a jenkins job, do you know a use-case for this ?

Also, Branch Specifier help tips only claim to support refs/heads/<branchName>, refs/remotes/<remoteRepoName>/<branchName> and refs/tags/<tagName>, so anything else to actually work would be only by chance

I'd like such a stricter approach you suggest, need to be carreful with legacy config conversion